### DISCUSSION TITLE
multiprocessing/joblib doesn't work in uvicorn

### DISCUSSION THREAD

**Discussion by qwertyforce** [Kludex/uvicorn#1340] [Q&A]
### Checklist

- [X] The bug is reproducible against the latest release or `master`.
- [X] There are no similar issues or pull requests to fix it yet.

### Describe the bug

app freezes when trying to use multiprocessing.Pool or joblib (https://github.com/joblib/joblib/issues/1250)

### Steps to reproduce the bug

```python
import uvicorn
import multiprocessing
if __name__ == '__main__':
    uvicorn.run('uvicorn_test:app', host='127.0.0.1', port=7777, log_level="info")

def calculate(func, args):
    result = func(*args)
    return '%s says that %s%s = %s' % (
        multiprocessing.current_process().name,
        func.__name__, args, result
        )

def calculatestar(args):
    return calculate(*args)

def mul(a, b):
    return a * b

def plus(a, b):
    return a + b

if __name__ == 'uvicorn_test':
    with multiprocessing.Pool(1) as pool:
        print(pool._processes)
        TASKS = [(mul, (i, 7)) for i in range(10)] + \
                [(plus, (i, 8)) for i in range(10)]
        imap_it = pool.imap(calculatestar, TASKS)
        for x in imap_it:
            print('\t', x)
```

### Expected behavior

print some text

### Actual behavior

app hangs

### Environment

Running uvicorn 0.17.0 with CPython 3.8.10 on Windows
joblib doesnt work on: linux, windows; python 3.7 3.8 3.9

**Comment by Kludex** [comment]
How do I run the example?

**Reply by qwertyforce** [reply]
@Kludex 
1. copy code into uvicorn_test.py
2. run
3. you will see this
![image](https://user-images.githubusercontent.com/44163887/151377772-2d3f9821-5d24-481e-af8d-267416a192e2.png)

4. comment out 
```python 
if __name__ == '__main__':
    uvicorn.run('uvicorn_test:app', host='127.0.0.1', port=7777, log_level="info")
```
in ```if __name__ == 'uvicorn_test':``` change uvicorn_test to ``` __main__ ```so code becomes like this
```python
import uvicorn
import multiprocessing
# if __name__ == '__main__':
#     uvicorn.run('uvicorn_test:app', host='127.0.0.1', port=7777, log_level="info")

def calculate(func, args):
    result = func(*args)
    return '%s says that %s%s = %s' % (
        multiprocessing.current_process().name,
        func.__name__, args, result
        )

def calculatestar(args):
    return calculate(*args)

def mul(a, b):
    return a * b

def plus(a, b):
    return a + b

if __name__ == '__main__':
    with multiprocessing.Pool(1) as pool:
        print(pool._processes)
        TASKS = [(mul, (i, 7)) for i in range(10)] + \
                [(plus, (i, 8)) for i in range(10)]
        imap_it = pool.imap(calculatestar, TASKS)
        for x in imap_it:
            print('\t', x)
```
5. run file again
![image](https://user-images.githubusercontent.com/44163887/151378387-f5e980c8-38f8-485f-8bbc-5a95705903af.png)
it's working

**Reply by Kludex** [reply]
Where's the app?

On Ubuntu 20.04:

```bash
❯ python uvicorn_test.py 
ERROR:    Error loading ASGI app. Attribute "app" not found in module "uvicorn_test".
```

**Reply by qwertyforce** [reply]
On Debian 10 it's working. I use uvicorn[standard]  0.17.0.post1.
idk you can add this to existing code
```python
from fastapi import FastAPI
app = FastAPI()
```
but it should work even without app.

**Comment by lovelydinosaur** [comment]
@Kludex - I think our general policy ought to be to push these kinda things into "discussions" first. (Probably?) Do you have permissions to do that ATM?

<img width="333" alt="Screenshot 2022-01-27 at 14 19 43" src="https://user-images.githubusercontent.com/647359/151377256-3ba92827-a5b4-4596-bf4c-b08e56de82d3.png">

**Reply by Kludex** [reply]
I do. Sure. :+1:
